src/libstd/process.rs

```diff
@@ -0,0 +1,837 @@
+use io::{self, Error, ErrorKind};
+use path::AsPath;
+use libc;
+use sync::mpsc::{channel, Receiver};
+use sys::fd::FileDesc;
+use sys::process2::Process as ProcessImp;
+use sys::process2::Command as CommandImp;
+use sys::process2::ExitStatus as ExitStatusImp;
+use sys;
+use sys_common::{AsInner, AsInnerMut};
+use thread::Thread;
+
+/// Representation of a running or exited child process.
+///
+/// This structure is used to represent and manage child processes. A child
+/// process is created via the `Command` struct, which configures the spawning
+/// process and can itself be constructed using a builder-style interface.
+///
+/// # Example
+///
+/// ```should_fail
+/// use std::old_io::Command;
+///
+/// let mut child = match Command::new("/bin/cat").arg("file.txt").spawn() {
+///     Ok(child) => child,
+///     Err(e) => panic!("failed to execute child: {}", e),
+/// };
+///
+/// let contents = child.stdout.as_mut().unwrap().read_to_end();
+/// assert!(child.wait().unwrap().success());
+/// ```
+pub struct Child {
+    handle: ProcessImp,
+
+    /// None until wait() or wait_with_output() is called.
+    status: Option<ExitStatusImp>,
+
+    /// The handle for writing to the child's stdin, if it has been captured
+    pub stdin: Option<ChildStdin>,
+
+    /// The handle for reading from the child's stdout, if it has been captured
+    pub stdout: Option<ChildStdout>,
+
+    /// The handle for reading from the child's stderr, if it has been captured
+    pub stderr: Option<ChildStderr>,
+}
+
+/// A handle to a child procesess's stdin
+pub struct ChildStdin {
+    inner: FileDesc
+}
+
+impl Write for ChildStdin {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.inner.write(buf)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
+
+/// A handle to a child procesess's stdout
+pub struct ChildStdout {
+    inner: FileDesc
+}
+
+impl Read for ChildStdout {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        self.inner.read(buf)
+    }
+}
+
+/// A handle to a child procesess's stderr
+pub struct ChildStderr {
+    inner: FileDesc
+}
+
+impl Read for ChildStderr {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        self.inner.read(buf)
+    }
+}
+
+/// The `Command` type acts as a process builder, providing fine-grained control
+/// over how a new process should be spawned. A default configuration can be
+/// generated using `Command::new(program)`, where `program` gives a path to the
+/// program to be executed. Additional builder methods allow the configuration
+/// to be changed (for example, by adding arguments) prior to spawning:
+///
+/// ```
+/// use std::old_io::Command;
+///
+/// let mut process = match Command::new("sh").arg("-c").arg("echo hello").spawn() {
+///   Ok(p) => p,
+///   Err(e) => panic!("failed to execute process: {}", e),
+/// };
+///
+/// let output = process.stdout.as_mut().unwrap().read_to_end();
+/// ```
+#[derive(Clone)]
```

Comment:
We may actually want to consider removing this `Clone` implementation. I've thought in the past that if `stdin`, `stdout`, and `stderr` happen to take some form of I/O object it will likely not implement `Clone`.

Although we could also just say that the `Clone` implementation clears the input/output handles.